File: frontend/src/services/proxmox.ts
import { apiClient } from "../utils/apiClient";

const API = "/api/v1/proxmox";

export interface ProxmoxNode {
    name: string;
    status: string;
    cpu_percent: number;
    memory_total_mb: number;
    memory_used_mb: number;
    disk_total_gb: number;
    disk_used_gb: number;
    uptime_seconds: number;
    version: string;
    ssl_fingerprint: string;
}

export interface ProxmoxVm {
    id: string;
    name: string;
    state: string;
    cpu_count: number;
    memory_assigned_mb: number;
    memory_startup_mb: number;
    memory_demand_mb: number;
    uptime_seconds: number;
    host_server: string;
    guest_os: string;
    creation_time: string | null;
    last_checkpoint: string | null;
    status_message: string;
    integration_services_enabled: boolean;
    cpu_usage_percent: number;
    disk_read_mbps: number;
    disk_write_mbps: number;
    network_receive_mbps: number;
    network_send_mbps: number;
}

export interface ProxmoxLxc {
    id: string;
    name: string;
    state: string;
    cpu_count: number;
    memory_assigned_mb: number;
    memory_startup_mb: number;
    memory_demand_mb: number;
    uptime_seconds: number;
    host_server: string;
    guest_os: string;
    creation_time: string | null;
    cpu_usage_percent: number;
    disk_read_mbps: number;
    disk_write_mbps: number;
    network_receive_mbps: number;
    network_send_mbps: number;
    status_message: string;
}

export interface ProxmoxNetwork {
    id: string;
    name: string;
    switch_type: string;
    vlan_id: number | null;
    status: string;
    node: string;
    cidr: string;
    address: string;
    gateway: string | null;
    connected_vms: number;
    type: string;
}

export interface ProxmoxStorage {
    id: string;
    name: string;
    path: string;
    size_bytes: number;
    used_bytes: number;
    type: string;
    status: string;
    content: string;
    node: string;
}

export interface ProxmoxTask {
    id: string;
    node: string;
    type: string;
    user: string;
    status: string;
    start_time: string | null;
    end_time: string | null;
    upid: string;
}

export interface ProxmoxSnapshot {
    id: string;
    name: string;
    vm_name: string;
    vm_id: string | null;
    checkpoint_type: string;
    creation_time: string | null;
    size_bytes: number;
    parent_checkpoint_id: string | null;
    notes: string | null;
}

export interface ProxmoxSummary {
    connected: boolean;
    cluster_name: string | null;
    version: string | null;
    nodes_online: number;
    nodes_total: number;
    total_vms: number;
    running: number;
    stopped: number;
    paused: number;
    total_lxc: number;
    running_lxc: number;
    stopped_lxc: number;
    total_cpu: number;
    total_memory_gb: number;
    used_memory_gb: number;
    total_storage_gb: number;
    used_storage_gb: number;
    storage_count: number;
    network_count: number;
    total_snapshots: number;
    error: string | null;
}

export interface ProxmoxHealthNode {
    name: string;
    status: string;
    cpu_percent: number;
    memory_percent: number;
    uptime_seconds: number;
    vm_count: number;
    version: string | null;
}

export interface ProxmoxHealth {
    connected: boolean;
    status: string;
    nodes: ProxmoxHealthNode[];
    cluster_summary: string | null;
    error: string | null;
}

export interface ProxmoxActionResponse {
    success: boolean;
    message: string | null;
    error: string | null;
    vm_name: string | null;
}

export const proxmoxApi = {
    async getSummary(): Promise<ProxmoxSummary> {
        return apiClient<ProxmoxSummary>(`${API}/overview`);
    },

    async getHealth(): Promise<ProxmoxHealth> {
        return apiClient<ProxmoxHealth>(`${API}/health`);
    },

    async testConnection(): Promise<{ connected: boolean; latency_ms: number; message: string | null; hostname: string | null; error: string | null }> {
        return apiClient(`${API}/test`);
    },

    async listNodes(): Promise<ProxmoxNode[]> {
        const data = await apiClient<{ items: ProxmoxNode[] }>(`${API}/nodes`);
        return data.items ?? [];
    },

    async listVms(): Promise<ProxmoxVm[]> {
        const data = await apiClient<{ items: ProxmoxVm[] }>(`${API}/vms`);
        return data.items ?? [];
    },

    async getVm(vmId: string): Promise<ProxmoxVm> {
        const data = await apiClient<{ item: ProxmoxVm }>(`${API}/vms/${vmId}`);
        return data.item;
    },

    async startVm(vmId: string): Promise<ProxmoxActionResponse> {
        return apiClient<ProxmoxActionResponse>(`${API}/vms/${vmId}/start`, { method: "POST" });
    },

    async stopVm(vmId: string, force = false): Promise<ProxmoxActionResponse> {
        return apiClient<ProxmoxActionResponse>(`${API}/vms/${vmId}/stop?force=${force}`, { method: "POST" });
    },

    async restartVm(vmId: string): Promise<ProxmoxActionResponse> {
        return apiClient<ProxmoxActionResponse>(`${API}/vms/${vmId}/restart`, { method: "POST" });
    },

    async pauseVm(vmId: string): Promise<ProxmoxActionResponse> {
        return apiClient<ProxmoxActionResponse>(`${API}/vms/${vmId}/pause`, { method: "POST" });
    },

    async resumeVm(vmId: string): Promise<ProxmoxActionResponse> {
        return apiClient<ProxmoxActionResponse>(`${API}/vms/${vmId}/resume`, { method: "POST" });
    },

    async listLxc(): Promise<ProxmoxLxc[]> {
        const data = await apiClient<{ items: ProxmoxLxc[] }>(`${API}/lxc`);
        return data.items ?? [];
    },

    async getLxc(vmId: string): Promise<ProxmoxLxc> {
        const data = await apiClient<{ item: ProxmoxLxc }>(`${API}/lxc/${vmId}`);
        return data.item;
    },

    async startLxc(vmId: string): Promise<ProxmoxActionResponse> {
        return apiClient<ProxmoxActionResponse>(`${API}/lxc/${vmId}/start`, { method: "POST" });
    },

    async stopLxc(vmId: string): Promise<ProxmoxActionResponse> {
        return apiClient<ProxmoxActionResponse>(`${API}/lxc/${vmId}/stop`, { method: "POST" });
    },

    async listNetworks(): Promise<ProxmoxNetwork[]> {
        const data = await apiClient<{ items: ProxmoxNetwork[] }>(`${API}/networks`);
        return data.items ?? [];
    },

    async listStorage(): Promise<ProxmoxStorage[]> {
        const data = await apiClient<{ items: ProxmoxStorage[] }>(`${API}/storage`);
        return data.items ?? [];
    },

    async listTasks(node?: string): Promise<ProxmoxTask[]> {
        const url = node ? `${API}/tasks?node=${node}` : `${API}/tasks`;
        const data = await apiClient<{ items: ProxmoxTask[] }>(url);
        return data.items ?? [];
    },

    async listSnapshots(vmId?: string): Promise<ProxmoxSnapshot[]> {
        const url = vmId ? `${API}/snapshots?vm_id=${vmId}` : `${API}/snapshots`;
        const data = await apiClient<{ items: ProxmoxSnapshot[] }>(url);
        return data.items ?? [];
    },

    async createSnapshot(vmId: string, name?: string): Promise<{ success: boolean; message: string | null; error: string | null }> {
        return apiClient(`${API}/snapshots`, {
            method: "POST",
            json: { vm_id: vmId, name },
        });
    },

    async deleteSnapshot(vmId: string, snapshotId: string): Promise<{ success: boolean; message: string | null; error: string | null }> {
        return apiClient(`${API}/vms/${vmId}/snapshots/${snapshotId}`, { method: "DELETE" });
    },
};

File: backend/app/schemas/proxmox.py
"""
Mission Control Proxmox Schemas

Pydantic request/response models for the Proxmox API.
"""

from pydantic import BaseModel

# ------------------------------------------------------------------ #
# VM                                                                  #
# ------------------------------------------------------------------ #


class ProxmoxVmResponse(BaseModel):
    id: str
    name: str
    state: str
    cpu_count: int
    memory_assigned_mb: int
    memory_startup_mb: int
    memory_demand_mb: int
    uptime_seconds: int
    host_server: str
    guest_os: str
    creation_time: str | None = None
    last_checkpoint: str | None = None
    status_message: str
    integration_services_enabled: bool
    cpu_usage_percent: float = 0.0
    disk_read_mbps: float = 0.0
    disk_write_mbps: float = 0.0
    network_receive_mbps: float = 0.0
    network_send_mbps: float = 0.0


class ProxmoxVmListResponse(BaseModel):
    count: int
    items: list[ProxmoxVmResponse]


class ProxmoxVmDetailResponse(BaseModel):
    item: ProxmoxVmResponse


class ProxmoxVmActionResponse(BaseModel):
    success: bool
    message: str | None = None
    error: str | None = None
    vm_name: str | None = None


# ------------------------------------------------------------------ #
# LXC Containers                                                      #
# ------------------------------------------------------------------ #


class ProxmoxLxcResponse(BaseModel):
    id: str
    name: str
    state: str
    cpu_count: int
    memory_assigned_mb: int
    memory_startup_mb: int
    memory_demand_mb: int
    uptime_seconds: int
    host_server: str
    guest_os: str
    creation_time: str | None = None
    cpu_usage_percent: float = 0.0
    disk_read_mbps: float = 0.0
    disk_write_mbps: float = 0.0
    network_receive_mbps: float = 0.0
    network_send_mbps: float = 0.0
    status_message: str


class ProxmoxLxcListResponse(BaseModel):
    count: int
    items: list[ProxmoxLxcResponse]


class ProxmoxLxcDetailResponse(BaseModel):
    item: ProxmoxLxcResponse


class ProxmoxLxcActionResponse(BaseModel):
    success: bool
    message: str | None = None
    error: str | None = None
    vm_name: str | None = None


class ProxmoxLxcTemplateResponse(BaseModel):
    id: str
    name: str
    file: str
    node: str
    size_bytes: int = 0
    os: str = ""
    description: str = ""
    version: str = ""
    arch: str = ""


class ProxmoxLxcTemplateListResponse(BaseModel):
    count: int
    items: list[ProxmoxLxcTemplateResponse]


class ProxmoxLxcCreateRequest(BaseModel):
    node: str
    vmid: str | None = None
    ostemplate: str
    hostname: str = "mission-control"
    cores: int = 2
    memory: int = 4096
    swap: int = 0
    disk: int = 8
    storage: str = "local-lvm"
    password: str | None = None
    unprivileged: bool = True
    nesting: bool = True
    net_bridge: str = "vmbr0"
    net_ip: str = "dhcp"
    nameserver: str | None = None
    searchdomain: str | None = None
    description: str | None = None


class ProxmoxLxcCreateResponse(BaseModel):
    success: bool
    vmid: str | None = None
    node: str | None = None
    message: str | None = None
    error: str | None = None


class ProxmoxLxcCloneRequest(BaseModel):
    new_vmid: str | None = None
    hostname: str | None = None


class ProxmoxLxcDeleteResponse(BaseModel):
    success: bool
    message: str | None = None
    error: str | None = None


# ------------------------------------------------------------------ #
# Node                                                                #
# ------------------------------------------------------------------ #


class ProxmoxNodeResponse(BaseModel):
    name: str
    status: str
    cpu_percent: float = 0.0
    memory_total_mb: int = 0
    memory_used_mb: int = 0
    disk_total_gb: float = 0.0
    disk_used_gb: float = 0.0
    uptime_seconds: int = 0
    version: str = ""
    ssl_fingerprint: str = ""


class ProxmoxNodeListResponse(BaseModel):
    count: int
    items: list[ProxmoxNodeResponse]


# ------------------------------------------------------------------ #
# Network                                                             #
# ------------------------------------------------------------------ #


class ProxmoxNetworkResponse(BaseModel):
    id: str
    name: str
    switch_type: str
    vlan_id: int | None = None
    status: str
    node: str
    cidr: str = ""
    address: str = ""
    gateway: str | None = None
    connected_vms: int = 0
    type: str = "unknown"


class ProxmoxNetworkListResponse(BaseModel):
    count: int
    items: list[ProxmoxNetworkResponse]


# ------------------------------------------------------------------ #
# Storage                                                             #
# ------------------------------------------------------------------ #


class ProxmoxStorageResponse(BaseModel):
    id: str
    name: str
    path: str
    size_bytes: int
    used_bytes: int
    type: str
    status: str = "available"
    content: str = ""
    node: str = ""


class ProxmoxStorageListResponse(BaseModel):
    count: int
    items: list[ProxmoxStorageResponse]


# ------------------------------------------------------------------ #
# Task                                                                #
# ------------------------------------------------------------------ #


class ProxmoxTaskResponse(BaseModel):
    id: str
    node: str
    type: str
    user: str
    status: str
    start_time: str | None = None
    end_time: str | None = None
    upid: str = ""


class ProxmoxTaskListResponse(BaseModel):
    count: int
    items: list[ProxmoxTaskResponse]


# ------------------------------------------------------------------ #
# Snapshot                                                            #
# ------------------------------------------------------------------ #


class ProxmoxSnapshotResponse(BaseModel):
    id: str
    name: str
    vm_name: str
    vm_id: str | None = None
    checkpoint_type: str = "snapshot"
    creation_time: str | None = None
    size_bytes: int = 0
    parent_checkpoint_id: str | None = None
    notes: str | None = None


class ProxmoxSnapshotListResponse(BaseModel):
    count: int
    items: list[ProxmoxSnapshotResponse]


class ProxmoxSnapshotCreateRequest(BaseModel):
    vm_id: str
    name: str | None = None


class ProxmoxSnapshotActionResponse(BaseModel):
    success: bool
    message: str | None = None
    error: str | None = None
    snapshot: dict | None = None


# ------------------------------------------------------------------ #
# Summary & Health                                                    #
# ------------------------------------------------------------------ #


class ProxmoxSummaryResponse(BaseModel):
    connected: bool
    cluster_name: str | None = None
    version: str | None = None
    nodes_online: int = 0
    nodes_total: int = 0
    total_vms: int = 0
    running: int = 0
    stopped: int = 0
    paused: int = 0
    total_lxc: int = 0
    running_lxc: int = 0
    stopped_lxc: int = 0
    total_cpu: int = 0
    total_memory_gb: float = 0.0
    used_memory_gb: float = 0.0
    total_storage_gb: float = 0.0
    used_storage_gb: float = 0.0
    storage_count: int = 0
    network_count: int = 0
    total_snapshots: int = 0
    error: str | None = None


class ProxmoxHealthNodeResponse(BaseModel):
    name: str
    status: str
    cpu_percent: float = 0.0
    memory_percent: float = 0.0
    uptime_seconds: int = 0
    vm_count: int = 0
    version: str | None = None


class ProxmoxHealthResponse(BaseModel):
    connected: bool
    status: str
    nodes: list[ProxmoxHealthNodeResponse] = []
    cluster_summary: str | None = None
    error: str | None = None


class ProxmoxConnectionTestResponse(BaseModel):
    connected: bool
    latency_ms: int = 0
    message: str | None = None
    version: str | None = None
    hostname: str | None = None
    error: str | None = None
